service/src/main_impl.rs

```diff
@@ -681,15 +730,62 @@ fn init_target_parentchain<E>(
 		.unwrap();
 }
 
-fn init_parentchain<E>(
+fn init_parentchain<E, ParentchainApi>(
 	enclave: &Arc<E>,
 	node_api: &ParentchainApi,
 	tee_account_id: &AccountId32,
 	parentchain_id: ParentchainId,
 ) -> (Arc<ParentchainHandler<ParentchainApi, E>>, Header)
+where
+	E: EnclaveBase + Sidechain,
+	ParentchainApi: ChainApi<Hash = Hash>
+		+ AccountApi
+		+ GetTransactionPayment
+		+ GetBalance
+		+ BalancesExtrinsics
+		+ SubmitAndWatch,
+{
+	let parentchain_handler = Arc::new(
+		ParentchainHandler::new_with_automatic_light_client_allocation(
+			node_api.clone(),
+			enclave.clone(),
+			parentchain_id,
+		)
+		.unwrap(),
+	);
+	let last_synced_header = parentchain_handler.init_parentchain_components().unwrap();
+	println!("[{:?}] last synced parentchain block: {}", parentchain_id, last_synced_header.number);
+
+	let nonce = node_api.get_nonce_of(tee_account_id).unwrap();
+	info!("[{:?}] Enclave nonce = {:?}", parentchain_id, nonce);
+	enclave.set_nonce(nonce, parentchain_id).unwrap_or_else(|_| {
+		panic!("[{:?}] Could not set nonce of enclave. Returning here...", parentchain_id)
+	});
+
+	let metadata = node_api.metadata().clone();
+	let runtime_spec_version = node_api.runtime_version().spec_version;
+	let runtime_transaction_version = node_api.runtime_version().transaction_version;
+	enclave
+		.set_node_metadata(
+			NodeMetadata::new(metadata, runtime_spec_version, runtime_transaction_version).encode(),
+			parentchain_id,
+		)
+		.unwrap_or_else(|_| {
+			panic!("[{:?}] Could not set the node metadata in the enclave", parentchain_id)
+		});
+
+	(parentchain_handler, last_synced_header)
+}
+
+fn init_integritee_parentchain<E>(
```

Comment:
this was just an attempt to solve the mess with redundant but explict code. not sure that's the way